src/mod_stream_management.erl

```diff
@@ -115,10 +202,88 @@ get_resume_timeout(Default) ->
 set_resume_timeout(ResumeTimeout) ->
     set_module_opt(?MYNAME, ?MODULE, resume_timeout, ResumeTimeout).
 
+
+-spec get_gc_repeat_after(pos_integer()) -> pos_integer().
+get_gc_repeat_after(Default) ->
+    S = gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, undefined),
+    case S of
+        {true, GCOpts} ->
+            proplists:get_value(gc_repeat_after, GCOpts, Default);
+        _ -> Default
+    end.
+
+-spec set_gc_repeat_after(pos_integer()) -> boolean().
+set_gc_repeat_after(ResumeTimeout) ->
+    S = gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, undefined),
+    case S of
+        {true, GCOpts} ->
+            NewGCOpts = lists:keystore(gc_repeat_after, 1, GCOpts,
+                                       {gc_repeat_after, ResumeTimeout}),
+            NewStaleOpts = {true, NewGCOpts},
+            set_module_opt(?MYNAME, ?MODULE, stale_h, NewStaleOpts);
+        _ -> false
+    end.
+
+-spec get_gc_geriatric(pos_integer()) -> pos_integer().
+get_gc_geriatric(Default) ->
+    S = gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, undefined),
+    case S of
+        {true, GCOpts} ->
+            proplists:get_value(gc_geriatric, GCOpts, Default);
+        _ -> Default
+    end.
+
+-spec set_gc_geriatric(pos_integer()) -> boolean().
+set_gc_geriatric(ResumeTimeout) ->
+    S = gen_mod:get_module_opt(?MYNAME, ?MODULE, stale_h, undefined),
+    case S of
+        {true, GCOpts} ->
+            NewGCOpts = lists:keystore(gc_geriatric, 1, GCOpts,
+                                       {gc_geriatric, ResumeTimeout}),
+            NewStaleOpts = {true, NewGCOpts},
+            set_module_opt(?MYNAME, ?MODULE, stale_h, NewStaleOpts);
+        _ -> false
+    end.
+
+
 %%
 %% API for `ejabberd_c2s'
 %%
 
+-spec make_smid() -> smid().
+make_smid() ->
+    base64:encode(crypto:strong_rand_bytes(21)).
```

Comment:
I took that function just as it was from `ejabberd_c2s`. You might notice on the file changes that that piece of code is now missing from `ejabberd_c2s` and was moved to `mod_stream_management`, precisely because I think it is the module responsible for generating these kind of things. Whether it is the best or not, well, it is just as it was. Maybe optimising/refactoring is out of the scope of this already super big PR 🤔

Reply:
About the `make_ref()`, not either, because this produced value is given to the client, and the client will save it and later reuse it when he requests resumption -> that's how resumption works, when the client requests so, he provides this "unique stream management identifier".

On the other hand, `base64:encode(erlang:ref_to_list(make_ref()))` does indeed generate a pretty usable random string. The problem is that it doesn't promise really unicity. Actually, `make_ref()` can reuse references if it can ensure that the previous one was garbage collected and not used any more, in which case, it's legit. Anyway, I don't know how that behaves when it comes to nodes on clusters and so on. We can think about that if we ever get a complain about `crypto:strong_rand_bytes(21)` duplicating keys 😛